feat: add entity metadata links

andrewthauer opened the pull request:
## Hey, I just made a Pull Request!

Partially addresses #3787

This introduces the base concepts for a new optional entity `metadata.links` field. Follow up PR's will introduce a common react catalog component to render entity links within entity pages.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

andrewthauer commented on packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts:
```diff
@@ -134,6 +134,12 @@ export class FieldFormatEntityPolicy implements EntityPolicy {
+    const links = entity.metadata.links ?? [];
+
+    for (let i = 0; i < links.length; ++i) {
+      optional(`links.${i}`, links[i]?.icon, this.validators.isValidEntityName);
```
Wasn't sure if I should introduce a new check or re-use the entity name validation for the `icon` key. I also wasn't sure if I should add new `required` validation check for the `url`.

freben replied:
Yeah there should for sure be a required check for that, but that's supposed to be covered by the `SchemaValidEntityPolicy`. So this new field should be added here with the proper semantics -> https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/SchemaValidEntityPolicy.ts#L21

andrewthauer replied:
Added additional validations and tests :)

Fox32 commented on packages/catalog-model/examples/acme/org.yaml:
```diff
@@ -3,6 +3,9 @@ kind: Group
 metadata:
   name: acme-corp
   description: The acme-corp organization
+  links:
```
Maybe we should have at least one example with more than one link? Except I missed it 😁

freben commented on packages/catalog-model/examples/apis/swapi-graphql.yaml:
```diff
@@ -3,1174 +3,13 @@ kind: API
-      cursor: String!
-    }
+  definition:
+    $text: https://github.com/graphql/swapi-graphql/blob/master/schema.graphql
```
Hm, this is an unrelated change - drawback being, now you're subject to github rate limits instead of all data being local.

I'm actually OK with it though. Should work fine as long as we have the implicit github.com integration always being added whether it's in one's config or not.

andrewthauer replied:
Good point. I'll revert this change.

Fox32 replied:
>Hm, this is an unrelated change - drawback being, now you're subject to github rate limits instead of all data being local.

In the end we are already doing it for most other API files, but I would prefer to keep this as-is to have another test case without `$text`.

freben replied:
Ah of course, you are right. But OK, revert sounds fine anyway

andrewthauer replied:
Sorry already reverted. Happy to change it back in another PR for another example.

freben commented on packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.test.ts:
```diff
@@ -110,4 +114,9 @@ describe('FieldFormatEntityPolicy', () => {
     data.metadata.tags.push('Hello World');
     await expect(policy.enforce(data)).rejects.toThrow(/tags.*"Hello World"/i);
   });
+
```
I'd like these tests to be very detailed and to exercise every single possible variation :)

- accepts missing links
- accepts empty links array
- rejects bad links value

- rejects missing links.url
- rejects empty links.url value
- rejects bad links.url value

- accepts missing links.label
- rejects empty links.label value
- rejects bad links.label value

- accepts missing links.icon
- rejects empty links.icon value
- rejects bad links.icon value

andrewthauer replied:
Done

freben commented on packages/catalog-model/src/entity/Entity.ts:
```diff
@@ -161,3 +166,23 @@ export type EntityRelationSpec = {
+/**
+ * A link to external information that is related to the entity.
+ */
+export type EntityLink = {
```
this should be exported in `index.ts` too

andrewthauer replied:
Added

freben commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -314,6 +314,34 @@ This field is optional, and currently has no special semantics.
 Each tag must be sequences of `[a-z0-9]` separated by `-`, at most 63 characters
 in total.
 
+### `links` [optional]
```
could you add this field to the example at the very top of the file? both in the json and yaml ones that are next to each other

freben commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -54,6 +54,11 @@ software catalog API.
+      "url": "https://admin.example-org.com",
+      "title": "Admin Dashboard",
+      "icon": "dashboard"
+    }],
```
in the yaml below too!

andrewthauer replied:
Whoops missed that. Added now.

andrewthauer commented:
@freben, @Fox32 - I believe I've addressed all the comments. I feel slightly bad about the `isValidUrl` function using a try catch on `new URL(url)`. This could be switched over to a well crafted regex, shared library, or some other suggestion.

andrewthauer commented on .changeset/fuzzy-pumpkins-tie.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/catalog-model': minor
```
Maybe this should be patch since it's additive?

Fox32 replied:
It's also not a breaking change, so patch should be fine?

freben commented on packages/catalog-model/src/validation/types.ts:
```diff
@@ -24,4 +24,7 @@ export type Validators = {
   isValidTag(value: unknown): boolean;
+  isValidLinkUrl(value: unknown): boolean;
+  isValidLinkTitle(value: unknown): boolean;
+  isValidLinkIcon(value: unknown): boolean;
```
This is an actual "breaking" change.

I'm not sure these are worth it. A valid url is a valid url - so having it in common is fine but maybe we can skip it here. And the title, the non-empty part of it is something we just typically slap in the schema because it makes absolutely no sense to allow the empty string. For the icons - well, not sure about that one, that's the only one where some marginal value could be had but, since it's just some nomenclature that people are free to choose as they see fit, and only within this very small scope of Backstage icons, I think the fixed rule of "same as metadata.name" is plenty forgiving.

Thoughts? I'd prefer just dropping these personally.

andrewthauer replied:
Yep, that seems to make sense to me. This is my first time making modifications to the catalog model so wasn't entirely sure what should and shouldn't be included.

freben replied:
Nah we don't. Could be improved wholesale across the project some time. We're switching to json-schema anyway so it'll be easy to slap checking regexes in there later 

https://github.com/backstage/backstage/pull/4277

(don't worry BTW about coordination, I'll let you know if we will collide, but I suspect you'll go in first anyway...)

freben replied:
Or... Did I read that question right? :)

andrewthauer replied:
Either way, #4277 doesn't seem to have many conflicts at all from my skim of the PR.

freben reviewed (approved):
A few little things remaining. Thank you for this work, almost there! <3

andrewthauer commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -81,6 +86,10 @@ metadata:
   tags:
     - java
+  links:
+    url: https://admin.example-org.com
```
Thanks for the catch 🤕

adamdmharvey commented on packages/catalog-model/examples/acme/org.yaml:
```diff
@@ -3,6 +3,11 @@ kind: Group
   description: The acme-corp organization
+  links:
+    - url: http://www.acme.com/
+      name: Website
```
Title, not name, ya?

andrewthauer replied:
😢 Where is my head today ...

adamdmharvey replied:
Pixies! https://www.youtube.com/watch?v=49FB9hhoO6c

adamdmharvey commented on packages/catalog-model/examples/acme/org.yaml:
```diff
@@ -3,6 +3,11 @@ kind: Group
+    - url: http://www.acme.com/
+      name: Website
+    - url: https://meta.wikimedia.org/wiki/
+      name: Intranet
```
Title, not name, ya?

andrewthauer commented on packages/catalog-model/src/validation/CommonValidatorFunctions.ts:
```diff
@@ -92,4 +92,35 @@ export class CommonValidatorFunctions {
+
+    try {
+      const url = new URL(value);
+      return !!url;
```
This was to work around linting errors. Eslint doesn't like `new URL(value)` (no-new violation). Assigning it to a variable and returning true means the variable isn't used 😞

freben replied:
I see. Anyway I think using the URL constructor like this to check was the right decision, doing it with a regex or so is just bound to be less robust.

freben replied:
Oh and by the way - since this was a perfectly legitimate use of a bare `new`, it would be fine to just silence the lint error with a comment as well.

adamdmharvey reviewed (approved):
Good re: my recommended changes, thanks @andrewthauer 👍
